### PR TITLE
Feat: TextField can scroll when disabled

### PR DESCRIPTION
**Pull request by bryanoli**
This PR is adding a flag parameter to the `TextField` widget. This flag controls whether the TextField ignores pointers. The flag takes priority over other TextField behaviors such as enabled, so it can be useful when trying to have a disabled TextField that can be scrolled (behavior observed using TextArea on the web).

Adding a flag parameter to `TextField` helps with more customization and flexibility to the widget which can improve user experience. I am open to other ideas.   

Fixes issue #140147 

Before: 

https://github.com/flutter/flutter/assets/66151079/293e5b4e-3126-4a00-824d-1530aeaa494b

After:

https://github.com/flutter/flutter/assets/66151079/08c1af09-3bf9-4b49-b684-dda4dd920503

Usage:
```dart
child: TextField(
  ignorePointer: false,
  enabled: false,
),
```

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/test/material/text_field_test.dart [hunk 1]
```diff
@@ -24,6 +24,7 @@ import 'package:flutter/rendering.dart';
 import 'package:flutter/services.dart';
 import 'package:flutter_test/flutter_test.dart';
 
+import '../cupertino/text_field_restoration_test.dart';
```
Comment (Renzo-Olivares) [inline packages/flutter/test/material/text_field_test.dart:27]: nit: I don't think this line is needed.

File: packages/flutter/lib/src/material/text_field.dart [hunk 2]
```diff
@@ -604,6 +605,10 @@ class TextField extends StatefulWidget {
   /// [InputDecoration.enabled] property.
   final bool? enabled;
 
+  /// If false the text field is "disabled" but still scrollable.
+  /// However, if [enabled] is true, then the text field is "active" and still scrollable.
+  final bool? ignorePointer;
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/material/text_field.dart:610]: Maybe instead of discussing the specific scrollable use-case this parameter is solving, you could talk about what it does in general. i.e. ignorePointer decides whether the widget will handle pointer events.

File: packages/flutter/lib/src/material/text_field.dart [hunk 3]
```diff
@@ -291,6 +291,7 @@ class TextField extends StatefulWidget {
     this.onAppPrivateCommand,
     this.inputFormatters,
     this.enabled,
+    this.ignorePointer,
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/material/text_field.dart:294]: nit: maybe `ignoresPointers`?

File: packages/flutter/lib/src/material/text_field.dart (hunk 2 above)
Comment (justinmc) [reply packages/flutter/lib/src/material/text_field.dart:610]: +1, especially since this parameter doesn't actually cause the field to be disabled like `enabled` does.

File: packages/flutter/lib/src/material/text_field.dart [hunk 4]
```diff
@@ -985,7 +990,7 @@ class _TextFieldState extends State<TextField> with RestorationMixin implements
   final GlobalKey<EditableTextState> editableTextKey = GlobalKey<EditableTextState>();
 
   @override
-  bool get selectionEnabled => widget.selectionEnabled;
+  bool get selectionEnabled => widget.selectionEnabled && _isEnabled;
```
Comment (justinmc) [inline packages/flutter/lib/src/material/text_field.dart:993]: Why was this change needed?

File: packages/flutter/test/material/text_field_test.dart [hunk 5]
```diff
@@ -6535,6 +6536,58 @@ void main() {
...
+    expect(firstPos.dy, lessThan(fourthPos.dy));
+    expect(inputBox.hitTest(BoxHitTestResult(), position: inputBox.globalToLocal(firstPos)), isTrue);
+    expect(inputBox.hitTest(BoxHitTestResult(), position: inputBox.globalToLocal(fourthPos)), isFalse);
+
+    final TestGesture gesture = await tester.startGesture(firstPos, pointer: 7);
+    await tester.pump();
+    await gesture.moveBy(const Offset(0.0, -1000.0));
+    await tester.pump(const Duration(seconds: 1));
+    // Wait and drag again to trigger https://github.com/flutter/flutter/issues/6329
+    // (No idea why this is necessary, but the bug wouldn't repro without it.)
```
Comment (justinmc) [inline packages/flutter/test/material/text_field_test.dart:6575]: Hmm we need to figure that out.  Maybe try `pumpAndSettle` for the previous one on line 6573?

File: packages/flutter/test/material/text_field_test.dart [hunk 6]
```diff
@@ -6535,6 +6536,58 @@ void main() {
...
+    await gesture.moveBy(const Offset(0.0, -1000.0));
+    await tester.pump(const Duration(seconds: 1));
+    await gesture.up();
+    await tester.pump();
+    await tester.pump(const Duration(seconds: 1));
+
+    // Now the first line is scrolled up, and the fourth line is visible.
+    final Offset newFirstPos = textOffsetToPosition(tester, kMoreThanFourLines.indexOf('First'));
+    final Offset newFourthPos = textOffsetToPosition(tester, kMoreThanFourLines.indexOf('Fourth'));
```
Comment (justinmc) [inline packages/flutter/test/material/text_field_test.dart:6584]: Maybe `finalFirstPos` etc.

**Review by justinmc — COMMENTED** [review]
The approach looks good, thanks for the PR!  I'm on board with adding this parameter since it seems that we can't match textarea's behavior without it.

File: packages/flutter/lib/src/material/text_field.dart (hunk 4 above)
Comment (bryanoli) [reply packages/flutter/lib/src/material/text_field.dart:993]: This change was made because the text selection menu (copy, paste, etc...) opens up when right-clicking on the textfield when disabled (and ignorePointers set to false). So the selectionEnabled should take into account for _isEnabled to prevent selection in all cases when the widget is disabled.

Comment (justinmc) [reply packages/flutter/lib/src/material/text_field.dart:995]: Ah makes sense, thanks for the explanation.

File: packages/flutter/lib/src/material/text_field.dart [hunk 7]
```diff
@@ -604,6 +605,12 @@ class TextField extends StatefulWidget {
   /// [InputDecoration.enabled] property.
   final bool? enabled;
 
+  /// Determines whether this widget ignores pointer events.
+  ///
+  /// If true, this widget does not respond to pointer events,
+  /// If false this widget responds to all pointer events.
+  final bool? ignorePointers;
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/material/text_field.dart:612]: nit: consider mentioning this is `null` by default.

**Review by Renzo-Olivares — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/material/text_field.dart (hunk 7 above)
Comment (justinmc) [reply packages/flutter/lib/src/material/text_field.dart:605]: +1 to this suggestion.

File: packages/flutter/test/material/text_field_test.dart [hunk 8]
```diff
@@ -6540,6 +6540,56 @@ void main() {
...
+    expect(firstPos.dy, lessThan(fourthPos.dy));
+    expect(inputBox.hitTest(BoxHitTestResult(), position: inputBox.globalToLocal(firstPos)), isTrue);
+    expect(inputBox.hitTest(BoxHitTestResult(), position: inputBox.globalToLocal(fourthPos)), isFalse);
+
+    final TestGesture gesture = await tester.startGesture(firstPos, pointer: 7);
+    await tester.pump();
+    await gesture.moveBy(const Offset(0.0, -1000.0));
+    await tester.pumpAndSettle();
+    await gesture.moveBy(const Offset(0.0, -1000.0));
+    await tester.pump(const Duration(seconds: 1));
```
Comment (justinmc) [inline packages/flutter/test/material/text_field_test.dart:6579]: Does this work with pumpAndSettle instead? Same question for the similar pump a few lines below.

**Review by justinmc — COMMENTED** [review]
The only thing that I think should be changed is the `tester.pump(const Duration(seconds: 1));`, mentioned below. Otherwise this looks good.

File: packages/flutter/test/material/text_field_test.dart (hunk 8 above)
Comment (bryanoli) [reply packages/flutter/test/material/text_field_test.dart:6579]: Yes, using pumpAndSettle does work.

File: packages/flutter/test/material/text_field_test.dart [hunk 9]
```diff
@@ -6535,6 +6535,56 @@ void main() {
...
+    expect(inputBox.hitTest(BoxHitTestResult(), position: inputBox.globalToLocal(fourthPos)), isFalse);
+
+    final TestGesture gesture = await tester.startGesture(firstPos, pointer: 7);
+    await tester.pump();
+    await gesture.moveBy(const Offset(0.0, -1000.0));
+    await tester.pumpAndSettle();
+    await gesture.moveBy(const Offset(0.0, -1000.0));
+    await tester.pumpAndSettle();
+    await gesture.up();
+    await tester.pump();
```
Comment (justinmc) [inline packages/flutter/test/material/text_field_test.dart:6576]: This pump is probably unnecessary.

**Review by justinmc — APPROVED** [review]
LGTM :+1: . Thanks for the fixes!

One last nit: don't forget [this comment](https://github.com/flutter/flutter/pull/140922/files#r1464009649) asking for a small change to the docs there.
